### PR TITLE
Subaru safety refactoring

### PR DESCRIPTION
**Pull request by martinl**
This PR fixes sign for subaru preglobal models desired_torque and adds driver torque limits from  https://github.com/commaai/openpilot/pull/1489
 
I also removed unused preglobal messages from SUBARU_L_TX_MSGS and updated fwd messages block list accordingly

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
Seems like subaru global and legacy don't share all that much. I think it would be cleaner to give legacy it's own tx and rx hooks instead of the conditions on nearly every check.

**Comment by martinl** [comment]
> Seems like subaru global and legacy don't share all that much. I think it would be cleaner to give legacy it's own tx and rx hooks instead of the conditions on nearly every check.

Thank you for the feedback, I'll implement the separate hooks for legacy platform. I think we would also need  some advice on how to implement legacy driver torque allowance and factor, I did not find a clean way to use different values in current tests

**Comment by martinl** [comment]
I did the legacy hooks separation refactoring, and added missing speed and brake checks for legacy. Currently 2 legacy tests still need fixing. I tested global safety with openpilot master and Crosstrek 2018 and it is working. I'll request testing with some legacy models after the 2 tests are fixed

**Comment by martinl** [comment]
Both failing tests are related to driver torque

0x371
torque_driver_new = (GET_BYTE(to_push, 3) >> 5) + (GET_BYTE(to_push, 4) << 3);
torque_driver_new = to_signed(torque_driver_new, 11);

`  SG_ Steer_Torque_Sensor : 29|11@1- (8,0) [-8500|8500] "" XXX`

0x164
int desired_torque = ((GET_BYTES_04(to_send) >> 8) & 0x1FFF);

` SG_ LKAS_Command : 8|13@1- (-1,0) [-4096|4096] "" XXX`

**Comment by martinl** [comment]
Forester 2017 Steer_Torque_Sensor and LKAS_Command signals

Screenshot 2020-05-20 at 23 54 51

File: board/safety/safety_subaru.h [hunk 1]
```diff
@@ -6,11 +6,13 @@ const uint32_t SUBARU_RT_INTERVAL = 250000;    // 250ms between real time checks
 const int SUBARU_MAX_RATE_UP = 50;
 const int SUBARU_MAX_RATE_DOWN = 70;
 const int SUBARU_DRIVER_TORQUE_ALLOWANCE = 60;
+const int SUBARU_L_DRIVER_TORQUE_ALLOWANCE = 600;
 const int SUBARU_DRIVER_TORQUE_FACTOR = 10;
+const int SUBARU_L_DRIVER_TORQUE_FACTOR = 1;
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:11]: Can we also separate the legacy constants into a separate block? same with the msg check structs

File: board/safety/safety_subaru.h [hunk 2]
```diff
@@ -219,18 +325,33 @@ static int subaru_fwd_hook(int bus_num, CAN_FIFOMailBox_TypeDef *to_fwd) {
...
+      }
+    }
+  }
+  // fallback to do not forward
+  return bus_fwd;
 }
 
-static void subaru_legacy_init(int16_t param) {
+
+static void subaru_init(int16_t param) {
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:351]: This is now the same as `nooutput_init`. Delete this and use that instead

File: board/safety/safety_subaru.h [hunk 3]
```diff
@@ -110,18 +97,77 @@ static int subaru_rx_hook(CAN_FIFOMailBox_TypeDef *to_push) {
...
+  valid = addr_safety_check(to_push, subaru_l_rx_checks, SUBARU_L_RX_CHECK_LEN,
+                            NULL, NULL, NULL);
+
+  bool unsafe_allow_gas = unsafe_mode & UNSAFE_DISABLE_DISENGAGE_ON_GAS;
+
+  if (valid && (GET_BUS(to_push) == 0)) {
+    int addr = GET_ADDR(to_push);
+    if (addr == 0x371) {
+      int torque_driver_new;
+      torque_driver_new = (GET_BYTE(to_push, 3) >> 5) + (GET_BYTE(to_push, 4) << 3);
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:127]: `(GET_BYTE(to_push, 4) << 3) | GET_BYTE(to_push, 3)`

**Comment by martinl** [comment]
this safety test fails after changing legacy safety to match dbc scaling values:
```
======================================================================
FAIL: test_against_torque_driver (test_subaru_legacy.TestSubaruLegacySafety)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/openpilot/panda/tests/safety/test_subaru_legacy.py", line 109, in test_against_torque_driver
    self.assertFalse(self._tx(self._torque_msg(-MAX_STEER)))
AssertionError: 1 is not false

----------------------------------------------------------------------
```

**Comment by martinl** [comment]
> this safety test fails after changing the dbc scaling factors:
> 
> ```
> ======================================================================
> FAIL: test_against_torque_driver (test_subaru_legacy.TestSubaruLegacySafety)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/tmp/openpilot/panda/tests/safety/test_subaru_legacy.py", line 109, in test_against_torque_driver
>     self.assertFalse(self._tx(self._torque_msg(-MAX_STEER)))
> AssertionError: 1 is not false
> 
> ----------------------------------------------------------------------
> ```

The tests were failing because driver torque scaling factor 8 caused rounding errors. We are currently testing a branch with scaling factor 1 and also increased driver torque factor to 10, to match global.

**Comment by adeebshihadeh** [comment]
If this is ready for review, can you fix the safety tests?

**Comment by martinl** [comment]
subaru legacy safety tests are configured to use the new subaru_outback_2015_generated dbc file from the opendbc PR 270, should I just rebase panda to latest master?

**Comment by adeebshihadeh** [comment]
I merged the opendbc PR, so rebasing and updating the commit in the Dockerfile should fix it.

**Comment by martinl** [comment]
thanks, all tests passed now

File: board/safety/safety_subaru.h [hunk 4]
```diff
@@ -53,36 +56,23 @@ static uint8_t subaru_compute_checksum(CAN_FIFOMailBox_TypeDef *to_push) {
...
 
   bool valid = false;
-  if (subaru_global) {
-    valid = addr_safety_check(to_push, subaru_rx_checks, SUBARU_RX_CHECK_LEN,
-                              subaru_get_checksum, subaru_compute_checksum, subaru_get_counter);
-  } else {
-    valid = addr_safety_check(to_push, subaru_l_rx_checks, SUBARU_L_RX_CHECK_LEN,
-                              NULL, NULL, NULL);
-  }
+  valid = addr_safety_check(to_push, subaru_rx_checks, SUBARU_RX_CHECK_LEN,
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:59]: should be one line

File: board/safety/safety_subaru.h [hunk 5]
```diff
@@ -110,18 +100,77 @@ static int subaru_rx_hook(CAN_FIFOMailBox_TypeDef *to_push) {
...
+      relay_malfunction_set();
+    }
+  }
+  return valid;
+}
+
+static int subaru_legacy_rx_hook(CAN_FIFOMailBox_TypeDef *to_push) {
+
+  bool valid = false;
+  valid = addr_safety_check(to_push, subaru_l_rx_checks, SUBARU_L_RX_CHECK_LEN,
```
Comment (adeebshihadeh) [inline board/safety/safety_subaru.h:121]: one line

File: board/safety/safety_subaru.h (hunk 4 above)
Comment (martinl) [reply board/safety/safety_subaru.h]: changed

File: board/safety/safety_subaru.h (hunk 5 above)
Comment (martinl) [reply board/safety/safety_subaru.h]: changed

**Pull request merged**